controllers/transaction.js

```diff
@@ -0,0 +1,86 @@
+const Transaction = require("../models/Transaction");
+const TransactionSlip = require("../models/TransactionSlip");
+
+// @desc    Update transaction
+// @route   PUT /api/v1/transactions/:id
+// @access  Private
+exports.updateTransaction = async (req, res, next) => {
+  try {
+    let transaction = await Transaction.findById(req.params.id);
+    // console.log(req.params.id);
+    //check transaction
+    if (!transaction) {
+      return res.status(404).json({
+        success: false,
+        message: `No transaction with the id of ${req.params.id}`,
+      });
+    }
+
+    // Status : [PENDING, COMPLETE, REJECTED, CANCELED]
+
+    //check status transaction is not COMPLETE and CANCELED
+    if (transaction.status === "COMPLETE" || transaction.status === "CANCELED") {
+      return res.status(404).json({
+        success: false,
+        message: `Transaction with the id of ${req.params.id}'s status is not available to update [Status: COMPLETE, Status: CANCELED]`,
+      });
+    }
+
+    //Check User or Admin
+
+    if (req.user.role === "admin") {
+      // ADMIN CASE: Admin can update when transaction was PENDING after user upload slip
+
+      //Check status of transaction
+      if (transaction.status !== "PENDING") {
+        return res.status(404).json({
+          success: false,
+          message: `Transaction with the id of ${req.params.id}'s status is not available to check [Status: REJECTED]`,
+        });
+      }
+
+    } else {
+      // USER CASE: When admin rejected transaction, USER will see transaction's status: PENDING 
+   
+      // Make sure user is the transaction owner
+      if (transaction.user.toString() !== req.user.id &&req.user.role !== "admin") {
+        return res.status(401).json({
+          success: false,
+          message: `User ${req.user.id} is not authorized to update this transaction`,
+        });
+      } 
+        
+      //Check status of transaction
+      if (transaction.status !== "REJECTED") {
+          return res.status(401).json({
+            success: false,
+            message: `Cannot update transaction with the id of ${req.params.id}'s status is not available to check [Status: PENDING]`,
+          });
+        }
+      }
+    
+    //Update value of transaction [Admin: update status, User: update slip]
+    transaction = await Transaction.findByIdAndUpdate(req.params.id, req.body, {
+      new: true,
+      runValidators: true,
+    });
+
+    //If NEW STATUS is COMPLETE
+    if(transaction.status === "COMPLETE") {
+      const transactionSlipId = transaction.submitted_slip_images[transaction.submitted_slip_images.length-1]
+      const transactionSlip = await TransactionSlip.findById(transactionSlipId);
+      transaction.successful_payment_date = transactionSlip.submit_time;
```

Comment:
If I'm not mistaken this would only change the local object attribute not the attribute on the database document?